File: tom_som_go_runtime/spec_serialization_order.go
package somruntime

// spec_serialization_order.go — model-aware member ordering for YAML
// serialization (AA1 criterion 7), a faithful port of
// `tom_som_dart_runtime/lib/src/spec_serialization_order.dart` (and
// `spec_serialization_order.js` / `.ts` / `.py`).
//
// A SpecDocument is a flat, path-keyed store; the native YAML codec normally
// emits keys alphabetically for clean diffs. Criterion 7 instead requires each
// class's members to be emitted in the order declared by their
// @SerializationOrder annotation (the SOM source declaration order).
//
// This helper turns a path into an ordinal tuple — the @SerializationOrder of
// each field crossed on the way down (plus the numeric sequence for a list
// item), mirroring the walk SpecReflection.Resolve performs. Comparing those
// tuples lexicographically reproduces a depth-first, member-order traversal:
// siblings of one class sort by their declared order, and the recursion carries
// that ordering down the tree. Form fields (which are sub-keys, not path
// segments) are ordered by their position in the owning @Form's field list.
//
// Unannotated members sort after annotated ones (fallback ordinal), then by
// path/name, so ordering is always total and deterministic.

import "sort"

// serializationUnorderedFallback is the ordinal used for members without a
// @SerializationOrder, so they sort after every annotated member while staying
// stable relative to each other.
const serializationUnorderedFallback = 1 << 30

// SpecSerializationOrder computes member ordering keys from a SpecModel.
type SpecSerializationOrder struct {
	refl *SpecReflection
}

// NewSpecSerializationOrder builds an ordering helper for the given model.
func NewSpecSerializationOrder(model *SpecModel) *SpecSerializationOrder {
	return &SpecSerializationOrder{refl: NewSpecReflection(model)}
}

// OrderKey returns the ordinal tuple for path: one entry per field crossed (its
// @SerializationOrder, or serializationUnorderedFallback), with a trailing
// entry for a list item's numeric sequence. A path that does not resolve yields
// an empty tuple (it then sorts by its string form only).
func (s *SpecSerializationOrder) OrderKey(path string) []int {
	segs := SpecPathSegments(path)
	if len(segs) == 0 || segs[0] == "" {
		return nil
	}
	root := s.refl.RootForSegment(segs[0])
	if root == nil {
		return nil
	}

	key := []int{}
	curClass := s.refl.ClassNamed(root.Type)
	for i := 1; i < len(segs); i++ {
		cls := curClass
		if cls == nil {
			break
		}
		seg := segs[i]

		field := s.matchField(cls, seg)
		if field != nil {
			key = append(key, orderOf(field))
			if field.Kind == SpecFieldKindComplex || field.Kind == SpecFieldKindSection {
				curClass = s.refl.ClassNamed(field.Type)
				continue
			}
			break // list container / leaf / form terminates the descent
		}

		// A list item segment: `<base>-<seq>`.
		split, ok := SplitListItemSegment(seg)
		if !ok {
			break
		}
		listField := s.matchField(cls, split.Base)
		if listField == nil || listField.Kind != SpecFieldKindList {
			break
		}
		key = append(key, orderOf(listField))
		key = append(key, split.Seq)
		if listField.ElementIsComplex {
			curClass = s.refl.ClassNamed(listField.ElementType)
			continue
		}
		break // scalar list item is a leaf
	}
	return key
}

// OrderPaths orders paths by their OrderKey (lexicographically), breaking ties
// by the path string so the result is a total, stable order.
func (s *SpecSerializationOrder) OrderPaths(paths []string) []string {
	items := make([]string, len(paths))
	copy(items, paths)
	keys := make(map[string][]int, len(items))
	for _, p := range items {
		keys[p] = s.OrderKey(p)
	}
	sort.SliceStable(items, func(i, j int) bool {
		return compareOrderKeys(keys[items[i]], items[i], keys[items[j]], items[j]) < 0
	})
	return items
}

// OrderFormFields orders the form-field names of the @Form at formPath by their
// declared position in the form's field list; names not found in the model sort
// after, alphabetically.
func (s *SpecSerializationOrder) OrderFormFields(formPath string, fieldNames []string) []string {
	resolution := s.refl.Resolve(formPath)
	var field *SpecField
	if resolution != nil {
		field = resolution.Field
	}
	positions := map[string]int{}
	if field != nil {
		for i, ff := range field.FormFields {
			positions[ff.Name] = i
		}
	}
	names := make([]string, len(fieldNames))
	copy(names, fieldNames)
	sort.SliceStable(names, func(i, j int) bool {
		a, b := names[i], names[j]
		pa, okA := positions[a]
		if !okA {
			pa = serializationUnorderedFallback
		}
		pb, okB := positions[b]
		if !okB {
			pb = serializationUnorderedFallback
		}
		if pa != pb {
			return pa < pb
		}
		return a < b
	})
	return names
}

func (s *SpecSerializationOrder) matchField(cls *SpecClass, segment string) *SpecField {
	for _, f := range cls.Fields {
		if s.refl.FieldSegment(f) == segment {
			return f
		}
	}
	return nil
}

func orderOf(field *SpecField) int {
	if field.SerializationOrder != nil {
		return *field.SerializationOrder
	}
	return serializationUnorderedFallback
}

// compareOrderKeys reproduces the Dart _compareKeys / JS _compareKeys total
// order: shorter/equal-prefix tuples order by length, then ties break on the
// path string.
func compareOrderKeys(a []int, aPath string, b []int, bPath string) int {
	n := len(a)
	if len(b) < n {
		n = len(b)
	}
	for i := 0; i < n; i++ {
		if a[i] != b[i] {
			if a[i] < b[i] {
				return -1
			}
			return 1
		}
	}
	if len(a) != len(b) {
		if len(a) < len(b) {
			return -1
		}
		return 1
	}
	if aPath < bPath {
		return -1
	}
	if aPath > bPath {
		return 1
	}
	return 0
}

File: tom_som_rust_runtime/src/spec_serialization_order.rs
//! `spec_serialization_order` — model-aware member ordering for YAML
//! serialization (AA1 criterion 7), a faithful port of the Go
//! `spec_serialization_order.go`.
//!
//! A [`SpecDocument`](crate::spec_document::SpecDocument) is a flat, path-keyed
//! store; the native YAML codec normally emits keys alphabetically for clean
//! diffs. Criterion 7 instead requires each class's members to be emitted in the
//! order declared by their `@SerializationOrder` annotation (the SOM source
//! declaration order).
//!
//! This helper turns a path into an ordinal tuple — the `@SerializationOrder` of
//! each field crossed on the way down (plus the numeric sequence for a list
//! item), mirroring the walk [`SpecReflection::resolve`] performs. Comparing
//! those tuples lexicographically reproduces a depth-first, member-order
//! traversal: siblings of one class sort by their declared order, and the
//! recursion carries that ordering down the tree. Form fields (which are
//! sub-keys, not path segments) are ordered by their position in the owning
//! `@Form`'s field list.
//!
//! Unannotated members sort after annotated ones (fallback ordinal), then by
//! path/name, so ordering is always total and deterministic.

use std::cmp::Ordering;

use crate::spec_model::{SpecClass, SpecField, SpecModel, SPEC_FIELD_KIND_COMPLEX,
    SPEC_FIELD_KIND_LIST, SPEC_FIELD_KIND_SECTION};
use crate::spec_paths::{spec_path_segments, split_list_item_segment};
use crate::spec_reflection::SpecReflection;

/// The ordinal used for members without a `@SerializationOrder`, so they sort
/// after every annotated member while staying stable relative to each other.
pub const SERIALIZATION_UNORDERED_FALLBACK: i64 = 1 << 30;

/// Computes member ordering keys from a [`SpecModel`].
pub struct SpecSerializationOrder<'a> {
    refl: SpecReflection<'a>,
}

impl<'a> SpecSerializationOrder<'a> {
    /// Builds an ordering helper for the given model.
    pub fn new(model: &'a SpecModel) -> SpecSerializationOrder<'a> {
        SpecSerializationOrder {
            refl: SpecReflection::new(model),
        }
    }

    /// Returns the ordinal tuple for `path`: one entry per field crossed (its
    /// `@SerializationOrder`, or [`SERIALIZATION_UNORDERED_FALLBACK`]), with a
    /// trailing entry for a list item's numeric sequence. A path that does not
    /// resolve yields an empty tuple (it then sorts by its string form only).
    pub fn order_key(&self, path: &str) -> Vec<i64> {
        let segs = spec_path_segments(path);
        if segs.is_empty() || segs[0].is_empty() {
            return Vec::new();
        }
        let root = match self.refl.root_for_segment(&segs[0]) {
            Some(r) => r,
            None => return Vec::new(),
        };

        let mut key: Vec<i64> = Vec::new();
        let mut cur_class = self.refl.class_named(&root.type_);
        let mut i = 1;
        while i < segs.len() {
            let cls = match cur_class {
                Some(c) => c,
                None => break,
            };
            let seg = &segs[i];

            if let Some(field) = self.match_field(cls, seg) {
                key.push(order_of(field));
                if field.kind == SPEC_FIELD_KIND_COMPLEX || field.kind == SPEC_FIELD_KIND_SECTION {
                    cur_class = self.refl.class_named(&field.type_);
                    i += 1;
                    continue;
                }
                break; // list container / leaf / form terminates the descent
            }

            // A list item segment: `<base>-<seq>`.
            let split = match split_list_item_segment(seg) {
                Some(s) => s,
                None => break,
            };
            let list_field = match self.match_field(cls, &split.base) {
                Some(f) if f.kind == SPEC_FIELD_KIND_LIST => f,
                _ => break,
            };
            key.push(order_of(list_field));
            key.push(split.seq);
            if list_field.element_is_complex {
                cur_class = self.refl.class_named(&list_field.element_type);
                i += 1;
                continue;
            }
            break; // scalar list item is a leaf
        }
        key
    }

    /// Orders `paths` by their [`SpecSerializationOrder::order_key`]
    /// (lexicographically), breaking ties by the path string so the result is a
    /// total, stable order.
    pub fn order_paths(&self, paths: &[String]) -> Vec<String> {
        let mut items: Vec<String> = paths.to_vec();
        let keys: Vec<(String, Vec<i64>)> =
            items.iter().map(|p| (p.clone(), self.order_key(p))).collect();
        items.sort_by(|a, b| {
            let ka = keys.iter().find(|(p, _)| p == a).map(|(_, k)| k).unwrap();
            let kb = keys.iter().find(|(p, _)| p == b).map(|(_, k)| k).unwrap();
            compare_order_keys(ka, a, kb, b)
        });
        items
    }

    /// Orders the form-field names of the `@Form` at `form_path` by their
    /// declared position in the form's field list; names not found in the model
    /// sort after, alphabetically.
    pub fn order_form_fields(&self, form_path: &str, field_names: &[String]) -> Vec<String> {
        let resolution = self.refl.resolve(form_path);
        let mut names: Vec<String> = field_names.to_vec();
        let positions: Vec<(String, usize)> = match resolution.and_then(|r| r.field) {
            Some(field) => field
                .form_fields
                .iter()
                .enumerate()
                .map(|(idx, ff)| (ff.name.clone(), idx))
                .collect(),
            None => Vec::new(),
        };
        let pos_of = |name: &str| -> usize {
            positions
                .iter()
                .find(|(n, _)| n == name)
                .map(|(_, i)| *i)
                .unwrap_or(SERIALIZATION_UNORDERED_FALLBACK as usize)
        };
        names.sort_by(|a, b| {
            let pa = pos_of(a);
            let pb = pos_of(b);
            if pa != pb {
                pa.cmp(&pb)
            } else {
                a.cmp(b)
            }
        });
        names
    }

    fn match_field<'c>(&self, cls: &'c SpecClass, segment: &str) -> Option<&'c SpecField> {
        cls.fields
            .iter()
            .find(|f| self.refl.field_segment(f) == segment)
    }
}

fn order_of(field: &SpecField) -> i64 {
    field.serialization_order.unwrap_or(SERIALIZATION_UNORDERED_FALLBACK)
}

/// Reproduces the Go `compareOrderKeys` total order: shorter/equal-prefix tuples
/// order by length, then ties break on the path string.
fn compare_order_keys(a: &[i64], a_path: &str, b: &[i64], b_path: &str) -> Ordering {
    let n = a.len().min(b.len());
    for i in 0..n {
        if a[i] != b[i] {
            return a[i].cmp(&b[i]);
        }
    }
    if a.len() != b.len() {
        return a.len().cmp(&b.len());
    }
    a_path.cmp(b_path)
}
